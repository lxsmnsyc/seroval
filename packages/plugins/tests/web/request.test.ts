import {
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serializeAsync,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSONAsync,
} from 'seroval';
import { describe, expect, it } from 'vitest';
import RequestPlugin from '../../web/request';

const EXAMPLE_URL = 'http://localhost:3000';
const EXAMPLE_BODY = 'Hello World!';

describe('Request', () => {
  describe('serializeAsync', () => {
    it('supports Request', async () => {
      const example = new Request(EXAMPLE_URL, {
        method: 'POST',
        body: EXAMPLE_BODY,
      });
      const result = await serializeAsync(example, {
        plugins: [RequestPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof example>(result);
      expect(back).toBeInstanceOf(Request);
      expect(await back.text()).toBe(await example.text());
      expect(back.url).toBe(example.url);
      expect(back.method).toBe(example.method);
    });

    it('supports already read Request', async () => {
      const example = new Request(EXAMPLE_URL, {
        method: 'POST',
        body: EXAMPLE_BODY,
      });
      await example.text();
      const result = await serializeAsync(example, {
        plugins: [RequestPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof example>(result);
      expect(back).toBeInstanceOf(Request);
      expect(back.body).toBe(null);
      expect(back.url).toBe(example.url);
      expect(back.method).toBe(example.method);
    });
  });
  describe('toJSONAsync', () => {
    it('supports Request', async () => {
      const example = new Request(EXAMPLE_URL, {
        method: 'POST',
        body: EXAMPLE_BODY,
      });
      const result = await toJSONAsync(example, {
        plugins: [RequestPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof example>(result, {
        plugins: [RequestPlugin],
      });
      expect(back).toBeInstanceOf(Request);
      expect(await back.text()).toBe(await example.text());
      expect(back.url).toBe(example.url);
      expect(back.method).toBe(example.method);
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports Request', async () => {
      const example = new Request(EXAMPLE_URL, {
        method: 'POST',
        body: EXAMPLE_BODY,
      });
      const result = await crossSerializeAsync(example, {
        plugins: [RequestPlugin],
      });
      expect(result).toMatchSnapshot();
    });

    it('supports already read Request', async () => {
      const example = new Request(EXAMPLE_URL, {
        method: 'POST',
        body: EXAMPLE_BODY,
      });
      const result = await crossSerializeAsync(example, {
        plugins: [RequestPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports Request', async () => {
        const example = new Request(EXAMPLE_URL, {
          method: 'POST',
          body: EXAMPLE_BODY,
        });
        const result = await crossSerializeAsync(example, {
          plugins: [RequestPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports Request', async () =>
      new Promise<void>((resolve, reject) => {
        const example = new Request(EXAMPLE_URL, {
          method: 'POST',
          body: EXAMPLE_BODY,
        });
        crossSerializeStream(example, {
          plugins: [RequestPlugin],
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    describe('scoped', () => {
      it('supports Request', async () =>
        new Promise<void>((resolve, reject) => {
          const example = new Request(EXAMPLE_URL, {
            method: 'POST',
            body: EXAMPLE_BODY,
          });
          crossSerializeStream(example, {
            plugins: [RequestPlugin],
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
    });
  });
  describe('toJSONAsync', () => {
    it('supports Request', async () => {
      const example = new Request(EXAMPLE_URL, {
        method: 'POST',
        body: EXAMPLE_BODY,
      });
      const result = await toCrossJSONAsync(example, {
        plugins: [RequestPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof example>(result, {
        plugins: [RequestPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(Request);
      expect(await back.text()).toBe(await example.text());
      expect(back.url).toBe(example.url);
      expect(back.method).toBe(example.method);
    });

    it('supports already read Request', async () => {
      const example = new Request(EXAMPLE_URL, {
        method: 'POST',
        body: EXAMPLE_BODY,
      });
      await example.text();
      const result = await toCrossJSONAsync(example, {
        plugins: [RequestPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof example>(result, {
        plugins: [RequestPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(Request);
      expect(back.body).toBe(null);
      expect(back.url).toBe(example.url);
      expect(back.method).toBe(example.method);
    });
  });
  describe('toCrossJSONStream', () => {
    it('supports Request', async () =>
      new Promise<void>((resolve, reject) => {
        const example = new Request(EXAMPLE_URL, {
          method: 'POST',
          body: EXAMPLE_BODY,
        });
        toCrossJSONStream(example, {
          plugins: [RequestPlugin],
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
  });
});
