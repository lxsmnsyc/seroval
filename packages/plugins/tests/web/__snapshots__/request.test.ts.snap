// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Request > crossSerializeAsync > scoped > supports Request 1`] = `"($R=>$R[0]=new Request("http://localhost:3000/",$R[1]={body:$R[2]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,cache:"default",credentials:"same-origin",headers:$R[3]=new Headers($R[4]=[$R[5]=["content-type","text/plain;charset=UTF-8"]]),integrity:"",keepalive:!1,method:"POST",mode:"cors",redirect:"follow",referrer:"about:client",referrerPolicy:""}))($R["example"])"`;

exports[`Request > crossSerializeAsync > supports Request 1`] = `"$R[0]=new Request("http://localhost:3000/",$R[1]={body:$R[2]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,cache:"default",credentials:"same-origin",headers:$R[3]=new Headers($R[4]=[$R[5]=["content-type","text/plain;charset=UTF-8"]]),integrity:"",keepalive:!1,method:"POST",mode:"cors",redirect:"follow",referrer:"about:client",referrerPolicy:""})"`;

exports[`Request > crossSerializeAsync > supports already read Request 1`] = `"$R[0]=new Request("http://localhost:3000/",$R[1]={body:$R[2]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,cache:"default",credentials:"same-origin",headers:$R[3]=new Headers($R[4]=[$R[5]=["content-type","text/plain;charset=UTF-8"]]),integrity:"",keepalive:!1,method:"POST",mode:"cors",redirect:"follow",referrer:"about:client",referrerPolicy:""})"`;

exports[`Request > crossSerializeStream > scoped > supports Request 1`] = `"($R=>$R[0]=new Request("http://localhost:3000/",$R[1]={body:$R[2]=($R[3]=d=>new ReadableStream({start:c=>{d.on({next:v=>{c.enqueue(v)},throw:v=>{c.error(v)},return:()=>{c.close()}})}}))($R[4]=($R[5]=(b,a,s,l,p,f,e,n)=>(b=[],a=!0,s=!1,l=[],p=0,f=(v,m,x)=>{for(x=0;x<p;x++)l[x]&&l[x][m](v)},n=(o,x,z,c)=>{for(x=0,z=b.length;x<z;x++)(c=b[x],(!a&&x===z-1)?o[s?"return":"throw"](c):o.next(c))},e=(o,t)=>(a&&(l[t=p++]=o),n(o),()=>{a&&(l[t]=void 0)}),{__SEROVAL_STREAM__:!0,on:o=>e(o),next:v=>{a&&(b.push(v),f(v,"next"))},throw:v=>{a&&(b.push(v),f(v,"throw"),a=s=!1,l.length=0)},return:v=>{a&&(b.push(v),f(v,"return"),a=!1,s=!0,l.length=0)}}))()),cache:"default",credentials:"same-origin",headers:$R[6]=new Headers($R[7]=[$R[8]=["content-type","text/plain;charset=UTF-8"]]),integrity:"",keepalive:!1,method:"POST",mode:"cors",redirect:"follow",referrer:"about:client",referrerPolicy:""}))($R["example"])"`;

exports[`Request > crossSerializeStream > scoped > supports Request 2`] = `"($R=>$R[4].next($R[9]=new Uint8Array($R[10]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,0,12)))($R["example"])"`;

exports[`Request > crossSerializeStream > scoped > supports Request 3`] = `"($R=>$R[4].return(void 0))($R["example"])"`;

exports[`Request > crossSerializeStream > supports Request 1`] = `"$R[0]=new Request("http://localhost:3000/",$R[1]={body:$R[2]=($R[3]=d=>new ReadableStream({start:c=>{d.on({next:v=>{c.enqueue(v)},throw:v=>{c.error(v)},return:()=>{c.close()}})}}))($R[4]=($R[5]=(b,a,s,l,p,f,e,n)=>(b=[],a=!0,s=!1,l=[],p=0,f=(v,m,x)=>{for(x=0;x<p;x++)l[x]&&l[x][m](v)},n=(o,x,z,c)=>{for(x=0,z=b.length;x<z;x++)(c=b[x],(!a&&x===z-1)?o[s?"return":"throw"](c):o.next(c))},e=(o,t)=>(a&&(l[t=p++]=o),n(o),()=>{a&&(l[t]=void 0)}),{__SEROVAL_STREAM__:!0,on:o=>e(o),next:v=>{a&&(b.push(v),f(v,"next"))},throw:v=>{a&&(b.push(v),f(v,"throw"),a=s=!1,l.length=0)},return:v=>{a&&(b.push(v),f(v,"return"),a=!1,s=!0,l.length=0)}}))()),cache:"default",credentials:"same-origin",headers:$R[6]=new Headers($R[7]=[$R[8]=["content-type","text/plain;charset=UTF-8"]]),integrity:"",keepalive:!1,method:"POST",mode:"cors",redirect:"follow",referrer:"about:client",referrerPolicy:""})"`;

exports[`Request > crossSerializeStream > supports Request 2`] = `"$R[4].next($R[9]=new Uint8Array($R[10]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,0,12))"`;

exports[`Request > crossSerializeStream > supports Request 3`] = `"$R[4].return(void 0)"`;

exports[`Request > serializeAsync > supports Request 1`] = `"new Request("http://localhost:3000/",{body:new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,cache:"default",credentials:"same-origin",headers:new Headers([["content-type","text/plain;charset=UTF-8"]]),integrity:"",keepalive:!1,method:"POST",mode:"cors",redirect:"follow",referrer:"about:client",referrerPolicy:""})"`;

exports[`Request > serializeAsync > supports already read Request 1`] = `"new Request("http://localhost:3000/",{body:null,cache:"default",credentials:"same-origin",headers:new Headers([["content-type","text/plain;charset=UTF-8"]]),integrity:"",keepalive:!1,method:"POST",mode:"cors",redirect:"follow",referrer:"about:client",referrerPolicy:""})"`;

exports[`Request > toCrossJSONStream > supports Request 1`] = `"{"t":25,"i":0,"s":{"url":{"t":1,"s":"http://localhost:3000/"},"options":{"t":10,"i":1,"p":{"k":["body","cache","credentials","headers","integrity","keepalive","method","mode","redirect","referrer","referrerPolicy"],"v":[{"t":25,"i":2,"s":{"factory":{"t":25,"i":3,"c":"seroval-plugins/web/ReadableStreamFactory"},"stream":{"t":31,"i":4,"a":[],"f":{"t":26,"i":5,"s":4}}},"c":"seroval/plugins/web/ReadableStream"},{"t":1,"s":"default"},{"t":1,"s":"same-origin"},{"t":25,"i":6,"s":{"t":9,"i":7,"l":1,"a":[{"t":9,"i":8,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain;charset=UTF-8"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},{"t":1,"s":""},{"t":2,"s":3},{"t":1,"s":"POST"},{"t":1,"s":"cors"},{"t":1,"s":"follow"},{"t":1,"s":"about:client"},{"t":1,"s":""}],"s":11},"o":0}},"c":"seroval-plugins/web/Request"}"`;

exports[`Request > toCrossJSONStream > supports Request 2`] = `"{"t":32,"i":4,"f":{"t":15,"i":9,"l":12,"c":"Uint8Array","f":{"t":19,"i":10,"s":[72,101,108,108,111,32,87,111,114,108,100,33]},"b":0}}"`;

exports[`Request > toCrossJSONStream > supports Request 3`] = `"{"t":34,"i":4,"f":{"t":2,"s":1}}"`;

exports[`Request > toJSONAsync > supports Request 1`] = `"{"t":{"t":25,"i":0,"s":{"url":{"t":1,"s":"http://localhost:3000/"},"options":{"t":10,"i":1,"p":{"k":["body","cache","credentials","headers","integrity","keepalive","method","mode","redirect","referrer","referrerPolicy"],"v":[{"t":19,"i":2,"s":[72,101,108,108,111,32,87,111,114,108,100,33]},{"t":1,"s":"default"},{"t":1,"s":"same-origin"},{"t":25,"i":3,"s":{"t":9,"i":4,"l":1,"a":[{"t":9,"i":5,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain;charset=UTF-8"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},{"t":1,"s":""},{"t":2,"s":3},{"t":1,"s":"POST"},{"t":1,"s":"cors"},{"t":1,"s":"follow"},{"t":1,"s":"about:client"},{"t":1,"s":""}],"s":11},"o":0}},"c":"seroval-plugins/web/Request"},"f":31,"m":[]}"`;

exports[`Request > toJSONAsync > supports Request 2`] = `"{"t":25,"i":0,"s":{"url":{"t":1,"s":"http://localhost:3000/"},"options":{"t":10,"i":1,"p":{"k":["body","cache","credentials","headers","integrity","keepalive","method","mode","redirect","referrer","referrerPolicy"],"v":[{"t":19,"i":2,"s":[72,101,108,108,111,32,87,111,114,108,100,33]},{"t":1,"s":"default"},{"t":1,"s":"same-origin"},{"t":25,"i":3,"s":{"t":9,"i":4,"l":1,"a":[{"t":9,"i":5,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain;charset=UTF-8"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},{"t":1,"s":""},{"t":2,"s":3},{"t":1,"s":"POST"},{"t":1,"s":"cors"},{"t":1,"s":"follow"},{"t":1,"s":"about:client"},{"t":1,"s":""}],"s":11},"o":0}},"c":"seroval-plugins/web/Request"}"`;

exports[`Request > toJSONAsync > supports already read Request 1`] = `"{"t":25,"i":0,"s":{"url":{"t":1,"s":"http://localhost:3000/"},"options":{"t":10,"i":1,"p":{"k":["body","cache","credentials","headers","integrity","keepalive","method","mode","redirect","referrer","referrerPolicy"],"v":[{"t":2,"s":0},{"t":1,"s":"default"},{"t":1,"s":"same-origin"},{"t":25,"i":2,"s":{"t":9,"i":3,"l":1,"a":[{"t":9,"i":4,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain;charset=UTF-8"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},{"t":1,"s":""},{"t":2,"s":3},{"t":1,"s":"POST"},{"t":1,"s":"cors"},{"t":1,"s":"follow"},{"t":1,"s":"about:client"},{"t":1,"s":""}],"s":11},"o":0}},"c":"seroval-plugins/web/Request"}"`;
